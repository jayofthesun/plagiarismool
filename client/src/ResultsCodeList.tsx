import { List, ListItem, ListItemText, WithStyles, withStyles } from "@material-ui/core";

import React from "react";
import { MainContext } from "./MainContext";
import './App.css';

/**
 * Props for the result code list
 * @member fileList the list of files the user uploaded
 * @member passIndexBack call for parent to make to pass index data back to update the viewed file
 */
interface IProps extends WithStyles<typeof styles> {
    fileList: File[];
    passIndexBack: (index: number) => any;
}

/**
 * states for the result code list
 */
interface IState {}

/**
 * Styles for this component
 */
const styles = {
    list: {
        backgroundColor: '#3f51b5',
        width: '20%',
        height: '80vh',
        overflow: 'scroll',
        borderRadius: '5px 0px 0px 5px',
        boxShadow: `0px 2px 1px -1px rgba(0,0,0,0.2), 
        0px 1px 1px 0px rgba(0,0,0,0.14), 0px 1px 3px 0px rgba(0,0,0,0.12)`,

    },
    listText: {
        fontSize: '13px',
        color: '#fff'
    },
    button: {
        padding: '10px'
    }
  };

/**
 * The component that displays a file explorer of the files uploaded by the user
 */
export class ResultsCodeList extends React.Component<IProps, IState> {
    static contextType = MainContext;

    /**
     * Updates the file index 
     * which the parent will use to update the file view
     * @param index the given file index
     */
    changeFileView(index: number) {
        this.props.passIndexBack(index);
    }

    render() {
        return <div className={this.props.classes.list}>
            <List disablePadding={true}>
                {this.props.fileList.map((file: File, index: number) => {
                    return (
                        <ListItem button key={file.name} className={this.props.classes.button} onClick={() => this.changeFileView(index)}>
                            <ListItemText classes={{ primary : this.props.classes.listText}} primary={file.name} />
                        </ListItem>
                    )
                })}
            </List>
        </div>
    }
} 

export default withStyles(styles)(ResultsCodeList)

