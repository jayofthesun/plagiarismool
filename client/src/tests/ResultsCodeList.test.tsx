import React from 'react';
import ResultsCodeList, { ResultsCodeList as OriginalList } from '../ResultsCodeList';
import { shallow } from 'enzyme';
import 'jest';
import { ListItem, ListItemText } from '@material-ui/core';
import { render } from '@testing-library/react';

Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

it('should render results code list correctly with given fields', () => {
    let file = new File([], "cool.ts");
    let file2 = new File([], "coolio.ts");

    const files = [file, file2];
    const component = render(<ResultsCodeList fileList={files} passIndexBack={() => { }} />);
    expect(component).toMatchSnapshot();
});

describe('tests for the result code list component', () => {
    //Dummy files
    const file = new File([], "cool.ts");
    const file2 = new File([], "coolio.ts");
    const file3 = new File([], 'sample.ts', {});

    //Test props
    var emptyProps = {
        fileList: [],
        passIndexBack: ((index: number) => undefined)
    };
    var oneListProps = {
        fileList: [file],
        passIndexBack: ((index: number) => undefined),
    };
    var multipleListProps = {
        fileList: [file, file2, file3],
        passIndexBack: ((index: number) => undefined),
    };

    it('should have no list items if given an empty list', () => {
        const wrapper = shallow(<ResultsCodeList {...emptyProps} />);
        const component = wrapper.find(OriginalList).dive();
        expect(component.find(ListItem), "no item list length").toHaveLength(0);
    });

    it('should have 1 item in the list if given a single item list', () => {
        const wrapper = shallow(<ResultsCodeList {...oneListProps} />);
        const component = wrapper.find(OriginalList).dive();
        expect(component.find(ListItem), "one item list has correct length").toHaveLength(1);
        expect(component.find(ListItemText).props().primary, "1st item in single list has correct name").toBe("cool.ts");
    });

    it('should have multiple items in the list if given a multiple item list', () => {
        const wrapper = shallow(<ResultsCodeList {...multipleListProps} />);
        const component = wrapper.find(OriginalList).dive();
        expect(component.find(ListItem), "multiple list length is correct").toHaveLength(3);
        expect(component.find(ListItemText).at(0).props().primary, "1st list item in multiple list should be correct").toBe("cool.ts");
        expect(component.find(ListItemText).at(1).props().primary, "2nd list item in multiple list should be correct").toBe("coolio.ts");
        expect(component.find(ListItemText).at(2).props().primary, "3rd list item in multiple list should be correct").toBe("sample.ts");
    });

    it('if you click on a highlighted line it will send index back to parent', () => {
        const updateLineCallBack = jest.fn((index: number) => { });
        const wrapper = shallow(<ResultsCodeList {...multipleListProps} passIndexBack={updateLineCallBack} />);
        const component = wrapper.find(OriginalList).dive();
        component.find(ListItem).at(1).simulate('click');
        expect(updateLineCallBack, "see if function to pass call back is called").toHaveBeenCalledTimes(1);
        expect(updateLineCallBack.mock.calls[0][0], "clicking on the second item should pass back correct index").toBe(1);
        component.find(ListItem).at(0).simulate('click');
        expect(updateLineCallBack, "see if function to pass call back is called again").toHaveBeenCalledTimes(2);
        expect(updateLineCallBack.mock.calls[1][0], "clicking on the first item should pass back correct index").toBe(0);
    });
});