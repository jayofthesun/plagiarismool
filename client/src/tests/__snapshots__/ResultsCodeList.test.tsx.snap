// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render results code list correctly with given fields 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ResultsCodeList-list-1"
      >
        <ul
          class="MuiList-root"
        >
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root ResultsCodeList-button-3 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary ResultsCodeList-listText-2 MuiTypography-body1 MuiTypography-displayBlock"
              >
                cool.ts
              </span>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
          <div
            aria-disabled="false"
            class="MuiButtonBase-root MuiListItem-root ResultsCodeList-button-3 MuiListItem-gutters MuiListItem-button"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiListItemText-primary ResultsCodeList-listText-2 MuiTypography-body1 MuiTypography-displayBlock"
              >
                coolio.ts
              </span>
            </div>
            <span
              class="MuiTouchRipple-root"
            />
          </div>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ResultsCodeList-list-1"
    >
      <ul
        class="MuiList-root"
      >
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root ResultsCodeList-button-3 MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary ResultsCodeList-listText-2 MuiTypography-body1 MuiTypography-displayBlock"
            >
              cool.ts
            </span>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
        <div
          aria-disabled="false"
          class="MuiButtonBase-root MuiListItem-root ResultsCodeList-button-3 MuiListItem-gutters MuiListItem-button"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemText-root"
          >
            <span
              class="MuiTypography-root MuiListItemText-primary ResultsCodeList-listText-2 MuiTypography-body1 MuiTypography-displayBlock"
            >
              coolio.ts
            </span>
          </div>
          <span
            class="MuiTouchRipple-root"
          />
        </div>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
